File: binary search/binary_search.cpp
#include <vector>


class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() -1;
        int m;


        while ( l <= r)
        {
            m = l + ((r -l) / 2);

            if (nums[m] > target)
            {
                r = m -1;
            }
            else if (nums[m] < target)
            {
                l = m +1;
            }
            else return m;
        }

        return -1;
    }
};

File: binary search/binary search.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        
        # wrong solution
        # hashMap = {}


        # for i,v in enumerate(nums):
        #     hashMap[v] = i

        # if target in nums:
        #     return hashMap[target]
        # else:
        #     return -1


        if target not in nums:
            return -1
        
        l = 0
        r = len(nums) -1


        while l <= r:
            m = (r - l) // 2

            if nums[m] > target:
                r = m -1
            elif nums[m] < target:
                l = m +1
            else:
                return m
            

        
        return -1

